flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -3222,4 +3222,29 @@ def fir_CUDAAllocateOp : fir_Op<"cuda_allocate", [AttrSizedOperandSegments,
   let hasVerifier = 1;
 }
 
+def fir_CUDADeallocateOp : fir_Op<"cuda_deallocate",
+    [MemoryEffects<[MemFree<DefaultResource>]>]> {
+  let summary = "Perform the device deallocation of data of an allocatable";
+
+  let description = [{
+    The fir.cuda_deallocate operation performs the deallocation on the device
+    of the data of an allocatable.
+  }];
+
+  let arguments = (ins Arg<AnyRefOrBoxType, "", [MemWrite]>:$box,
```

Comment:
Technically it will have both a read and write effect (same for allocate actually): the runtime will/may read info from the incoming descriptor in both cases. I guess the memory effect would anyway prevent optimizers to think they can skip previous writes if this is the only user of box left, but probably better to add it.

Reply:
That's a good observation. I'll update the effects

Reply:
Done.